addons/marketing_campaign/models/marketing_campaign.py

```diff
@@ -0,0 +1,122 @@
+        partner_field = self.partner_field_id.name
+        if partner_field:
+            return getattr(record, partner_field)
+        elif self.object_id.model == 'res.partner':
+            return record
+
+    @api.multi
+    def state_running_set(self):
+        # TODO check that all subcampaigns are running
+        self.ensure_one()
+
+        if not self.activity_ids:
+            raise UserError(_("The campaign cannot be started. There are no activities in it."))
+
+        has_start = False
+        has_signal_without_from = False
+
+        for activity in self.activity_ids:
+            if activity.start:
+                has_start = True
+            if activity.signal and len(activity.from_ids) == 0:
+                has_signal_without_from = True
+
+        if not has_start and not has_signal_without_from:
+            raise UserError(_("The campaign cannot be started. It does not have any starting activity. Modify campaign's activities to mark one as the starting point."))
+
+        return self.write({'state': 'running'})
+
+    @api.multi
+    def state_done_set(self):
+        # TODO check that this campaign is not a subcampaign in running mode.
+        segments = self.env['marketing.campaign.segment'].search(
+            [('campaign_id', 'in', self.ids), ('state', '=', 'running')])
+        if segments:
+            raise UserError(_("The campaign cannot be marked as done before all segments are closed."))
+        return self.write({'state': 'done'})
+
+    @api.multi
+    def state_cancel_set(self):
+        # TODO check that this campaign is not a subcampaign in running mode.
+        return self.write({'state': 'cancelled'})
+
+    # prevent duplication until the server properly duplicates several levels of nested o2m
+    _sql_constraints = [
+        ('campaign_uniq', 'unique (name)', 'Duplicating campaigns is not supported.')
+    ]
+
+    def _find_duplicate_workitems(self, record, campaign_rec):
+        """Finds possible duplicates workitems for a record in this campaign, based on a uniqueness
+           field.
+           :param record: browse_record to find duplicates workitems for.
+           :param campaign_rec: browse_record of campaign
+        """
+        Workitems = self.env['marketing.campaign.workitem']
+        duplicate_workitem_domain = [('res_id','=', record.id),
+                                     ('campaign_id','=', campaign_rec.id)]
+        unique_field = campaign_rec.unique_field_id
+        if unique_field:
+            unique_value = getattr(record, unique_field.name, None)
+            if unique_value:
+                if unique_field.ttype == 'many2one':
+                    unique_value = unique_value.id
+                similar_res = self.env[campaign_rec.object_id.model].search(
+                    [(unique_field.name, '=', unique_value)])
+                if similar_res:
+                    duplicate_workitem_domain = [('res_id','in', similar_res),
```

Comment:
this will obviously through AssertionError: Invalid value.... similar_res is recordset not ids